### ISSUE TITLE
Static library builds?

### ISSUE THREAD

**Issue by hs-apotell** [copperspice/copperspice#161]
Is there a plan to support static library builds? Would this even be something that would still conform to the licensing agreement?

Reasoning - We are using a very small number of features and yet are generating over 300 MB of final packaged build (of which over 70% is CS binaries and PDBs). We would like to strip all the unnecessary/unused code and employing the linker to do the heavy lifting would be ideal.

**Comment by bgeller** [comment]
There are two issues related to static libraries for CopperSpice. One involves rewriting part of the plugin system to allow statically linking plugins. The other issue, which may be minor, your application must then be released as LGPL.  

In terms of making your files smaller our first suggestion is to do a Release build and disable the PDB files from MSVC.  I would also check your optimization settings.

**Comment by hs-apotell** [comment]
> One involves rewriting part of the plugin system to allow statically linking plugins.

Is this on the roadmap?

> your application must then be released as LGPL

We are open to providing object files and instructions to accompany that would satisfy the license requirements.

> do a Release build and disable the PDB files from MSVC

Unfortunately, this is not an option since we are building a middleware and not the final product itself. Our clientele expects debugging information to go along with the libraries. We are exploring the option to host a symbol server that an engineer can point to. Regardless, reducing the file sizes would be a long term benefit.

**Comment by agserm** [comment]
> > One involves rewriting part of the plugin system to allow statically linking plugins.
>
> Is this on the roadmap?

It is on our roadmap, however it is not currently a high priority. We have consulting resources and this is an area you may want to consider a support subscription.
 
> > your application must then be released as LGPL
>
> We are open to providing object files and instructions to accompany that would satisfy the license requirements.

It sounds like you have a good understanding of the LGPL requirements, so this is not an issue.
 
> > do a Release build and disable the PDB files from MSVC
> 
> Unfortunately, this is not an option since we are building a middleware and not the final product itself. Our clientele expects debugging information to go along with the libraries. We are exploring the option to host a symbol server that an engineer can point to. Regardless, reducing the file sizes would be a long term benefit.

I did not realize that your use case actually requires distributing debugging symbols. It is possible that distributing stripped PDBs for CopperSpice and full PDBs for your code will support the level of debugging your clients need.

**Comment by hs-apotell** [comment]
Does CopperSpice provide prebuilt binaries for supported platforms? I am looking specifically for Windows, Ubuntu and Mac. I wanted to compare the built binary sizes on different platforms. Here are the numbers from my local build on Ubuntu + gcc9. Could someone please confirm if these sizes look _reasonably_ close to what's expected.

```
                       Debug                RelWithDebInfo             Release
======================================================================================
libCsCore.so |  328806512 (313.57 MB) |  910278824 (868.11 MB) | 12638392 (12.05 MB)
libCsGui.so  | 1846930624   (1.72 GB) | 8361356096   (7.79 GB) | 60103312 (57.32 MB)
```

The RelWithDebInfo binaries sizes seem out of order. I expected them to be somewhere close to what Debug binaries are but they are almost 4x larger.

Are the CMakeLists.txt configured to build for RelWithDebInfo?

**Comment by janwilmans** [comment]
Binary builds are available from https://download.copperspice.com/copperspice/binary/cs-1.7/
The size of the RelWithDebInfo can be smaller due to optimizations, a lot less symbols will be present.

**Comment by hs-apotell** [comment]
Aren't Debug builds suppose to be bigger than _any_ variant of Release? But here the RelWithDebInfo are 4x the size of Debug. This is impacting downstream builds too. For one of our internal libraries, the Debug build is 803301854 bytes (766.09 MB) but the RelWithDebInfo is 1583823976 bytes (1.48 GB) - that's about 2x bigger. This is true for Windows platform as well. 
```
                       Debug                      RelWithDebInfo
==========================================================================
CsCore.dll |  22.2 MB (23,299,072 bytes)  |  12.9 MB ( 13,547,008 bytes)
CsCore.pdb |  88.1 MB (92,434,432 bytes)  | 106.0 MB (111,702,016 bytes)

CsGui.dll  |  97.2 MB (101,928,448 bytes) |  50.8 MB ( 53,367,296 bytes)
CsGui.pdb  | 387.0 MB (406,769,664 bytes) | 450.0 MB (472,838,144 bytes)
==========================================================================
```

I will spend more time investigating the root cause (possibly /LTCG or /GL option being turned on).

**Comment by agserm** [comment]
We will investigate this as well, any input you can provide will be helpful.

**Comment by gvanem** [comment]
>  But here the RelWithDebInfo are 4x the size of Debug. 

Maybe the use of the `cl /bigobj` switch has something to do with it? AFAICS, it's *not* needed for `src/core/*.cpp`. 
But certainly for  `src/gui/*.cpp`.

**Comment by agserm** [comment]
We did some research, and we have found no indication that this would be the cause of the file size issue. The bigobj flag does expand the indexes into the section table from 16-bit to 32-bit but that can not explain a 4x increase in overall file size.

Although CsCore does not technically need the bigobj flag, our initial testing indicated it was more reliable to use the same setting for all libraries. We could reevaluate this.
